### ISSUE TITLE
[css-grid][css-contain] Clarify that `contain:size` affects track sizing

### ISSUE THREAD

**Issue by MatsPalmgren** [w3c/csswg-drafts#4931]
Consider the following case:
```html
<div style="display:inline-grid; contain:size;">
  <item style="background:red">Hello Kitty</item>
</div>
```

My understanding of the CSS Containment spec is that the intent of `contain:size` is that all size calculations should be done as if the element has no children:
https://drafts.csswg.org/css-contain/#containment-size
Chromium and Gecko agrees that the content-box size in the above case is 0x0, which is correct. I tend to think that the track sizes also should be zero and thus that it should render the same as:
```html
<div style="display:inline-grid; grid:0/0;">
  <item style="background:red">Hello Kitty</item>
</div>
```
Neither Chromium nor Gecko does that currently, but I intend to fix this in Gecko ([bug 1488878](https://bugzilla.mozilla.org/show_bug.cgi?id=1488878)).

I can't find any spec text dealing with this case explicitly though, so I would like CSS Grid to clarify that  when the Grid container has `contain:size` the [Track Sizing Algorithm](https://drafts.csswg.org/css-grid/#algo-track-sizing) should be run as if the container has no items. This will have the effect that any [intrinsic track sizes](https://drafts.csswg.org/css-grid/#intrinsic-sizing-function) will be zero.

**Comment by MatsPalmgren** [comment]
(I've filed a Chromium [bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1069329).)

**Comment by bfgeek** [comment]
I don't believe that is the intent of the containment specification. From the spec:

> When calculating the size of the containing box, including when computing its intrinsic size, it must be treated as having no contents.

This is effectively just an early out when computing the intrinsic sizes for an element.

> Then, its contents must then be laid out into the containing box's resolved size.

Any layout algorithm should not then change its behaviour after the above step. E.g. these two cases should render the same:

```html
<div style="display:inline-grid; contain:size;">
  <item style="background:red">Hello Kitty</item>
</div>
```

```html
<div style="display:inline-grid; width: 0px;">
  <item style="background:red">Hello Kitty</item>
</div>
```

**Comment by Loirooriol** [comment]
Chromium and Firefox are not interoperable in these cases:

```html
<div style="display: inline-grid; grid-auto-columns: 50px; height: 50px; contain: size; border: 5px solid">
  <div style="background: yellow"></div>
</div>
```

```html
<div style="display: inline-grid; grid-template-columns: 50px; height: 50px; contain: size; border: 5px solid"></div>
```

In Chromium, the content area of the grid container is 50px wide, in Firefox it's 0px.

@MatsPalmgren, in https://bugzilla.mozilla.org/show_bug.cgi?id=1488878#c1 you say that what you want to implement is

1. place items into the grid as usual
2. run Track Sizing algo but pretend we have no items
3. calculate the container's size from the track sizes as usual

This seems what Chromium is doing, i.e. step 1 can create implicit tracks, tracks may end up being bigger than 0px in step 2, and this affects the size of the grid container in step 3.

However, according to the CSS Grid spec, once we know the size of the grid container, it's laid out again "for real". So there are more steps:

4. place items into the grid as usual
5. run Track Sizing algo, this time taking the items into account

Then I'm not sure why you think that the tracks should be 0px in your example. Do you want to avoid laying the grid again if there is size containment?

**Comment by MatsPalmgren** [comment]
> > When calculating the size of the containing box, including when computing its intrinsic size, it must be treated as having no contents.
> 
> This is effectively just an early out when computing the intrinsic sizes for an element.

The "_including_ when computing its intrinsic size" (my emphasis) seems to imply that it affects more size calculations than just the intrinsic size of the box itself. I don't see anything in the spec that supports the idea that it _only_ affects the intrinsic box size.
You can think of the calculation of [_intrinsic_ track sizes](https://drafts.csswg.org/css-grid/#intrinsic-sizing-function) as an "intrinsic sizing calculation" if that helps since it's based on the intrinsic sizes of the items (the chapter is even named [Resolve Intrinsic Track Sizes](https://drafts.csswg.org/css-grid/#algo-content)).

> `<div style="display:inline-grid; width: 0px;">`

Did you forget the `height:0`? The [`contain:size`](https://drafts.csswg.org/css-contain/#valdef-contain-size) spec is pretty clear that the children shouldn't affect the size: "This ensures that the containing box can be laid out without needing to examine its descendants." so clearly you don't want the container to have its height sized by the item.

Anyway, the Containment spec already [says](https://drafts.csswg.org/css-contain/#containment-size) that "this does not necessarily make the element zero-sized: properties set on the element itself, such as the columns property or the grid property, continue to be taken into account.". I read "taken into account" there as "run the TSA pretending that the grid container has no items", which is what I suggested in the OP. I don't think there is a way to take the grid properties into account in any other way.
 IOW, what I'm suggesting is merely making concrete what the spec is already suggesting as I see it.

**Comment by frivoal** [comment]
My understanding is the same as @bfgeek and @Loirooriol here. The "inculding intrinsic size" bit means that if the layout of any parent of the size-contained element depends in some way on the intrinsic size of that size-contained element, then that input to its layout is the same as if that size-contained element was empty.

For instance, the size of a grid row depends on the intrinsic sizes of the elements in that row. Say you have a column which is sized to min-content, we look at the intrinsic min content size of the elements in the column, pick the largest, size the column that way, and then do the actual layout of the grid items in that column at that width. During that phase, if any of the grid items have size containment, they'll be treated as empty for the purpose of figuring out their min-content size.

However, setting size-containment on the grid itself would not affect the intrinsic size calculation of the grid items.

> The "*including* when computing its intrinsic size" (my emphasis) seems to imply that it affects more size calculations than just the intrinsic size of the box itself.

No, it means that it's both the size of the box in actual layout, and the intrinsic size of the box if requested by its parent, that are calculated as if the box were empty. Not that other unspecified operations (such as track sizing) also treat it as empty. The spec tries to be explicit that once you've done this sizing step, everything else runs normally, with the following statement:

> Then, its contents must then be laid out into the containing box's resolved size.

----

> Anyway, the Containment spec already says that "this does not necessarily make the element zero-sized: properties set on the element itself, such as the columns property or the grid property, continue to be taken into account.". I read "taken into account" there as "run the TSA pretending that the grid container has no items", which is what I suggested in the OP. I don't think there is a way to take the grid properties into account in any other way.

For the purpose of finding how big the grid is, yes. Hence the grid in your example being sized to `width:0; height:0;` But then we do actual layout of the grid, and at that point everything behaves as usual, and so it's track aren't empty.

At least:
* that's what I meant when I phrased the spec that way
* doing what you suggest instead doesn't seem necessary to allow the kind of optimizations that size-containment is trying to enable. (to be fair, I don't think it would disable them either.)

That said, if it is not clear (which seems to be the case), the spec should certainly at least be clarified. If we have agreement that the behavior suggested here would be more useful, I suspect we could switch to that, but it's not clear to me that it is.

----

> Chromium and Firefox are not interoperable in these cases:
> 
> ```html
> <div style="display: inline-grid; grid-auto-columns: 50px; height: 50px; contain: size; border: > 5px solid">
>   <div style="background: yellow"></div>
> </div>
> ```
>
> In Chromium, the content area of the grid container is 50px wide, in Firefox it's 0px.

Yeah, in that case, Chromium is right and Firefox is wrong. That is exactly what the note in the spec is about:
> Even when the element’s sizing properties are auto this does not necessarily make the element zero-sized: properties set on the element itself, such as the columns property or the grid property, continue to be taken into account.

**Comment by frivoal** [comment]
> Chromium and Firefox are not interoperable in these cases […]

WPT has (at least) two tests for that, which Firefox fails:
https://wpt.fyi/results/css/css-contain/contain-size-grid-002.html
https://wpt.fyi/results/css/css-contain/contain-size-grid-003.html

Interestingly, Firefox does pass the corresponding multicol tests:
https://wpt.fyi/results/css/css-contain/contain-size-multicol-001.html
https://wpt.fyi/results/css/css-contain/contain-size-multicol-as-flex-item.html

**Comment by MatsPalmgren** [comment]
> > The "_including_ when computing its intrinsic size" (my emphasis) seems to imply that it affects more size calculations than just the intrinsic size of the box itself.
> 
> No, it means that it's both the size of the box in actual layout, and the intrinsic size of the box if requested by its parent, that are calculated as if the box were empty. Not that other unspecified operations (such as track sizing) also treat it as empty.

This is a fundamental misunderstanding of how Grid layout works. Track sizing isn't "other unspecified operations" unrelated to a grid container's size. It's literally the way you calculate its content-box size, [per spec](https://drafts.csswg.org/css-grid/#intrinsic-sizes): "_The max-content size (min-content size) of a grid container is the sum of the grid container’s track sizes (including gutters) in the appropriate axis, when the grid is sized under a max-content constraint (min-content constraint)._" So suggesting that size-containment doesn't affect track sizing makes no sense because if it doesn't then it can't affect the grid container's size either.

To be concrete, [here's](https://searchfox.org/mozilla-central/rev/9120151ddb35f2d4c37bfe613a54a4f10a9a3dc5/layout/generic/nsGridContainerFrame.cpp#8012,8038,8053-8054,8059,8061) how Gecko reports "the intrinsic size of the box if requested by its parent". I've highlighted the three salient steps there:
1. place items into the grid
2. run the Track Sizing Algorithm (TSA) in the relevant axis
3. return the sum of the track sizes + gaps

> The spec tries to be explicit that once you've done this sizing step, everything else runs normally, with the following statement:
> 
> > Then, its contents must then be laid out into the containing box's resolved size.

OK, so it seems what you're trying to say then is that `contain:size` only has an effect when the grid container is sized under a min-/max-content constraint, and has no effect otherwise?

> > [...] I read "taken into account" there as "run the TSA pretending that the grid container has no items", which is what I suggested in the OP. I don't think there is a way to take the grid properties into account in any other way.
> 
> For the purpose of finding how big the grid is, yes. Hence the grid in your example being sized to `width:0; height:0;` But then we do actual layout of the grid, and at that point everything behaves as usual, and so it's track aren't empty.

I'm guessing "how big the grid is" means "what the content-box size of the grid container is". (please try to use precise terms that are defined in specs or we're going to talk past each other).

>  doing what you suggest instead doesn't seem necessary to allow the kind of optimizations that size-containment is trying to enable.

Doing what you suggest means that we have to run the TSA twice: once _without_ items to determine the grid container's content-box size and once _with_ items to determine the track sizes to use for layout of the children. Always running the TSA without items, as I suggested in the OP, is an optimization (and a lot simpler to implement since running it once is what we're currently doing).

> Yeah, in that case, Chromium is right and Firefox is wrong. That is exactly what the note in the spec is about:
> 
> > Even when the element’s sizing properties are auto this does not necessarily make the element zero-sized: properties set on the element itself, such as the columns property or the grid property, continue to be taken into account.

Right, but the way you "take the grid property into account" is to run the TSA. So again, there is no other way to implement this but **to run the TSA without items** (at least under a min-/max-content constraint) and use the track sizes that falls out. If you're thinking of something else then please point to spec text that defines that alternative algorithm.

I can live with doing it twice I guess, if you all think it's important that intrinsic track sizes should be sized normally before laying out the children.

**Comment by Loirooriol** [comment]
@MatsPalmgren Can you clarify what's your proposal if you don't want to run the TSA without items when sizing the grid container under a min/max-content constraint, and then run the TSA with items when laying it out for real?

In your first comment you say that the track size should be 0, seemingly implying that you only want to run the TSA without items, never with items.

In your last comment you seem to mean that you want to ignore `grid` (so that you can avoid running the TSA) when sizing the grid container under a min/max-content constraint. And then run the TSA once with items when laying out for real.

So not sure I understand your proposal.

**Comment by MatsPalmgren** [comment]
> In your first comment you say that the track size should be 0, seemingly implying that you only want to run the TSA without items, never with items.

Right, that's option A.

> In your last comment you seem to mean that you want to ignore `grid` (so that you can avoid running the TSA) when sizing the grid container under a min/max-content constraint. And then run the TSA once with items when laying out for real.

No, I never said `grid` should be ignored. On the contrary, running the TSA is how you take `grid` into account. You can run it without items though. Then definite track sizes are still sized normally, but intrinsic track sizes will be zero. So option B (that I described in my last comment) is to run the TSA once _without_ items for calculating the intrinsic content-box size of the grid container, and then a second time _with_ items when "laying it out for real" (so that the final track sizes are sized normally as if `contain:size` weren't specified). The grid container's intrinsic size would still be taken from the first step though (the first step isn't needed if it has a definite size obviously). So for example:
```html
<div style="display:inline-grid; contain:size; grid:auto 100px / auto 200px">
  <item><div style="width:300px; height:300px"></div></item>
</div>
```
with option A:
the grid container content-box size is `200px x 100px`
the resolved track sizes are `0px 100px / 0px 200px`
the item's content-box size is `0px x 0px`
with option B:
the grid container content-box size is `200px x 100px`
the resolved track sizes are `300px 100px / 300px 200px`
the item's content-box size is `300px x 300px`

> So not sure I understand your proposal.

I've described two different options (so far). I hope they are clear now.

BTW, I'm assuming above that the [grid item placement step](https://drafts.csswg.org/css-grid/#placement) in the Grid layout algorithm _always_ occurs, also for `contain:size`.  But this isn't at all clear from the [CSS Containment spec](https://drafts.csswg.org/css-contain/#size-containment). In fact, a strict reading of the spec implies that **it shouldn't occur**: "_When calculating the size of the containing box, including when computing its intrinsic size, it must be treated as having no contents._"

The "_treated as having no contents_" implies that implicit tracks should **not** be created (since there are no items if we treat it as having no contents), for example:
```html
<div style="display:inline-grid; contain:size; grid-auto-rows:100px; grid-auto-columns:100px">
  <item></item>
</div>
```
Chrome makes the grid container content-box size `100px x 100px`, i.e. clearly not treating it as empty when computing its intrinsic size, which contradicts what the spec text currently says.

So the spec needs to say explicitly which parts of Grid layout are supposed to be run with/without items, unless it means all of them.

I don't have a strong opinion on which option to choose. Option B seems to be close to what Chrome is currently doing and that seems to be the behavior you guys prefer, so that's fine with me.  It needs to be clearly specified how it's actually supposed to work though, because the current spec text suggests something completely different.

**Comment by MatsPalmgren** [comment]
Also, the part about honoring `grid` etc is mentioned in [a Note](https://drafts.csswg.org/css-contain/#size-containment) in the spec. The [Conformance section](https://drafts.csswg.org/css-contain/#conformance) says that notes aren't normative if I read it correctly: "_All of the text of this specification is normative except sections explicitly marked as non-normative, examples, and notes._"

I assume honoring `grid` is supposed to be a normative requirement and thus it shouldn't (only) be mentioned in a note.

**Comment by frivoal** [comment]
The [definition of intrinsic sizes](https://drafts.csswg.org/css-sizing-3/#intrinsic-size) is explicitly about the min-content size and max-content size, not about track sizes or any other sizable aspect of an box (grid track sizes, multicol column sizes, etc). So by my reading (and by my intent when writing it), the spec calls for finding out the min-content width and height and max-content width and height (if you ever need them) and the width and height of the element as if it were empty, and once you've done that, you do full layout using these sizes. Yes, it does imply that you run the grid track sizing algorithm twice, once with empty tracks (for sizing purposes) and once with actual content (when doing the layout). A possible implementation strategy could be to have a specialized (and simpler/faster) implementation of the track sizing algo that only does track-sizing of an empty grid.

I'm still not understanding how the spec can be seen as ambiguous about that. For example, you said:
> In fact, a strict reading of the spec implies that it shouldn't occur: "When calculating the size of the containing box, including when computing its intrinsic size, it must be treated as having no contents."
> 
> The "treated as having no contents" implies that implicit tracks should not be created (since there are no items if we treat it as having no contents)

However, to get to this interpretation, you need to ignore the "when calculating …" part of the sentence. *When you calculate the size of the box*, you treat it as having no content, and therefore do not reserve room for any implicit track. *When laying out the grid into that size*, you do the full grid layout, with its content, including sizing the tracks correctly, and creating implicit tracks as necessary.

Even if it is not ambiguous, it may still be a bad idea. I don't think it is a bad idea, but we can discuss that.

* The advantage of switching to "option A" would be that we would only need to size the tracks once, rather than twice. Even with specialized first pass that can only handle empty grids but works faster on those, doing things once is should always be faster than doing them twice.

* The disadvantage of doing "option A"  is that it makes the feature less useful: gird sizes its track the way it does because that's the most useful thing to do for the layout of the grid. While contain:size defines what happens to auto-sized elements, it's largely meant to be used on elements that have a fixed author-supplied size. And regardless of how we end up getting the size we ant to layout into, within that size, calculating the track sizes normally will result in a normal grid layout. Calculating track sizes as if the grid were empty and laying out the grid items within those will result in a worse layout, and that difference is not necessary to realize the benefits expected from size containment.


Anyway, If I understand correctly, you're not necessarily advocating for "option A", merely saying that the spec could be read that way, and want it to be clearer on whether it means A or B.

I remain unconvinced that it can be read to mean option A, but nonetheless, if you have a suggestion of better phrasing to gets unambiguously to option B, it's always better to be clearer.

By the way, the reason why the grid text is in a note is that no additional normative requirement is meant on grid, and this is just an illustration/reminder of how the preceding normative text is meant to be applied to grid. I believe that option A would be special rules for grid, and would need to be normative.

**Comment by Loirooriol** [comment]
> with option A: [...] the item's content-box size is 0px x 0px

The grid area would be 0x0, yes, but wouldn't the item be bigger due to the automatic minimum size? Or would it not apply since we would kind of be treating the `auto` track sizing function as `0px`? On the one hand, I think it's bad if the automatic minimum size forces an item to grow bigger than its grid area. On the other hand, the item being 0x0 also seems bad. So I'm not a fan of option A.

> When you calculate the size of the box, you treat it as having no content, and therefore do not reserve room for any implicit track.

@frivoal You don't? So Chromium is wrong?

To summarize this issue, below I list the possibly unclear questions about the interaction of grid and size containment.

When sizing a grid container with size containment under a min/max-content constraint,

1. Should we run the placement algorithm (creating implicit tracks if needed)?
   Florian says we shouldn't? Chromium does, Firefox does not.
2. Should we run the track sizing algorithm?
   The spec says so. Chromium does, Firefox just sets the size of the grid container to 0.
3. If 1 and 2 are affirmative, should tracks take intrinsic contributions of the items into account?
   No, that would defeat the point of using size containment.

When laying out the grid container with size containment into that size,

4. Should we run the placement algorithm (creating implicit tracks if needed)?
   Yes, we need to know where to place the items in order to display them.
5. Should we run the track sizing algorithm?
   Yes, we need to know the sizes of the grid areas in order to size the grid items.
6. If 4 and 5 are affirmative, should tracks take intrinsic contributions of the items into account?
   Yes with option B, not with option A. Chromium and Firefox do it.
7. If 6 is negative, should the automatic minimum size still apply?
   Maybe not?

**Comment by frivoal** [comment]
> >    When you calculate the size of the box, you treat it as having no content, and therefore do not reserve room for any implicit track.
>
> @frivoal You don't? So Chromium is wrong?

I think so:
* spec based reasoning: spec says to size the element as if it were empty. implicit tracks with a non-zero fixed min track sizing function would have an impact on the size, and only exist if the element is not empty. Therefore, don't take them into account when calculating the size. (but then when later doing the layout pass, sure, do consider the grid items and do create the relevant tracks).
* goal based reasoning: if we take the number of elements into account in order to figure out how many implicit tracks we have, and these implicit tracks have a non-zero fixed min track sizing function, then, by applying display:none to grid items, we can change the size of the grid. This would mean that the element is not actually contained, and optimizations based on containment would become incorrect.

So, answers to your numbered questions are:

> When sizing a grid container with size containment under a min/max-content constraint,
> 
> 1. Should we run the placement algorithm (creating implicit tracks if needed)?
>    Florian says we shouldn't? Chromium does, Firefox does not.

No, that would defeat the point of using size containment.

> 2. Should we run the track sizing algorithm?
>    The spec says so. Chromium does, Firefox just sets the size of the grid container to 0.

Yes. We could switch to no without defeating the point of using size containment, but that would result in a less useful layout.

> 3. If 1 and 2 are affirmative, should tracks take intrinsic contributions of the items into account?
>    No, that would defeat the point of using size containment.

Agreed.

> When laying out the grid container with size containment into that size,
> 
> 4. Should we run the placement algorithm (creating implicit tracks if needed)?
>    Yes, we need to know where to place the items in order to display them.

Agreed

> 5. Should we run the track sizing algorithm?
>    Yes, we need to know the sizes of the grid areas in order to size the grid items.

Agreed. We could switch to no without defeating the point of using size containment, but that would result in a less useful layout.


> 6. If 4 and 5 are affirmative, should tracks take intrinsic contributions of the items into account?
>    Yes with option B, not with option A. Chromium and Firefox do it.

Yes. We could switch to no without defeating the point of using size containment, but that would result in a less useful layout. Also, if authors have a layout which would be resilient to that and want the extra optimization opportunities, they can apply size containment to the grid items.


> 7. If 6 is negative, should the automatic minimum size still apply?
>    Maybe not?

I don't think 6 should be negative, so ¯\\\_(ツ)\_/¯

**Comment by MatsPalmgren** [comment]
> The [definition of intrinsic sizes](https://drafts.csswg.org/css-sizing-3/#intrinsic-size) is explicitly about the min-content size and max-content size, not about track sizes or any other sizable aspect of an box (grid track sizes, multicol column sizes, etc).

But to calculate the intrinsic size of a grid container you need to calculate its track sizes first, since the [Grid spec](https://drafts.csswg.org/css-grid/#intrinsic-sizes) literally says that its intrinsic size is derived from those.  Hence the question of how they should be calculated for `contain:size` arise.

> So by my reading (and by my intent when writing it), the spec calls for finding out the min-content width and height and max-content width and height (if you ever need them) and the width and height of the element as if it were empty, and once you've done that, you do full layout using these sizes.

But the [contain:size](https://drafts.csswg.org/css-contain/#containment-size) spec says that it should be treated as empty in _more cases_ than just for intrinsic sizing though:
"_When calculating the size of the containing box, including when computing its intrinsic size, it must be treated as having no contents._"

If you meant that it should only affect the intrinsic size of the container's box then you need to change it to "_When calculating the intrinsic size of the containing box, it must be treated as having no contents._" so that it excludes the "when laying out the grid into that size" phase (to use your phrasing).

The ambiguity with Grid layout though is that the track sizes are calculated _again_ during the layout phase and that part of layout is considered an [intrinsic sizing step](https://drafts.csswg.org/css-grid/#algo-content) (not the container's intrinsic size, but the track sizes). So I think the Containment spec needs to make it clear that that step is excluded from the "treat the container as empty" condition.

> Anyway, If I understand correctly, you're not necessarily advocating for "option A", merely saying that the spec could be read that way, and want it to be clearer on whether it means A or B.

Correct, I claim that the `contain:size` spec is ambiguous (for Grid at least). I haven't advocated for any particular behavior (so far).

> By the way, the reason why the grid text is in a note is that no additional normative requirement is meant on grid, and this is just an illustration/reminder of how the preceding normative text is meant to be applied to grid.

Right, but the problem is that, for a few of the possible interpretations of the spec, it's not just clarifying the preceding text. If we go for the "it should be treated as empty so implicit tracks aren't created" option then yes, in that case it's only a clarifying note. Otherwise, it's not just clarifying but instead modifies/contradicts the preceding text and thus shouldn't be in a note.

**Comment by MatsPalmgren** [comment]
I think I agree with your answers to @Loirooriol's questions but let me clarify one point to avoid any misunderstanding:

> > When sizing a grid container with size containment under a min/max-content constraint,
> > 
> > 1. Should we run the placement algorithm (creating implicit tracks if needed)?
> >    Florian says we shouldn't? Chromium does, Firefox does not.
> 
> No, that would defeat the point of using size containment.

This is the wrong way to see it IMO.  We _should_ run the placement algorithm, and all other algorithms for that matter, it's just that when we treat the container as empty then we have zero items and thus we don't create any implicit tracks. IOW, _when calculating the container's intrinsic size_ for `<container style="...; contain:size">X</container>` it's **always** calculated identically to `<container style="...; contain:size"></container>` regardless of what the `...` expands to.

So the spec changes I would suggest are:

1. change "_When calculating the size of the containing box, including when computing its intrinsic size, it must be treated as having no contents._" to "_When calculating the intrinsic size of the containing box, it must be treated as having no contents._". Perhaps also add a parenthetical clarification: "_(including any pseudo-elements)_" at the end.

2. remove the note that follows that text since it's now redundant and just adds FUD

3. Change "_Then, its contents must then be laid out into the containing box's resolved size._" to "_Then, its contents is laid out into the containing box's resolved size normally, i.e. `contain:size` has no effect in this phase._"

(3 is needed to clarify that the [Resolve Intrinsic Track Sizes](https://drafts.csswg.org/css-grid/#algo-content) step for Grid should not treat the container as empty during layout. (Perhaps there are other cases (multicol? flexbox?) where this also matters, I haven't checked.))

**Comment by MatsPalmgren** [comment]
BTW, is "_containing box_" really a good term to use? Isn't "_container's box_" clearer? Or perhaps "[_principal box_](https://drafts.csswg.org/css-display/#principal-box)" would be better since it's an already established term.

**Comment by Loirooriol** [comment]
> spec based reasoning: spec says to size the element as if it were empty. implicit tracks with a non-zero fixed min track sizing function would have an impact on the size, and only exist if the element is not empty. Therefore, don't take them into account when calculating the size.

@frivoal Well, but the spec says "When calculating the size of the containing box". The [Grid Item Placement Algorithm](https://drafts.csswg.org/css-grid/#auto-placement-algo) is not part of the [Grid Sizing Algorithm](https://drafts.csswg.org/css-grid/#algo-overview), it's a previous step. So if the size containment is supposed to apply not only during sizing but also during previous steps that can affect sizing, then the spec should probably clarify it.

>  if we take the number of elements into account in order to figure out how many implicit tracks we have, and these implicit tracks have a non-zero fixed min track sizing function, then, by applying display:none to grid items, we can change the size of the grid. This would mean that the element is not actually contained

It seems to me this would be more like style containment rather than size containment.

My understanding is that size containment is only supposed to prevent the intrinsic size of an element from depending on the size of the contents. But in this case, the existence of implicit tracks wouldn't depend on the layout of the children, only on their computed style.

So I'm not opposed to avoid creating implicit tracks with size containment, but I'm not convinced that's necessary.

> This is the wrong way to see it IMO. We should run the placement algorithm

@MatsPalmgren But in practice it doesn't matter, right? The placement algorithm places items, if it behaves as having no contents then there are no items to place and the placement algorithm is no-op.

**Comment by frivoal** [comment]
> 3. Change "_Then, its contents must then be laid out into the containing box's resolved size._" to "_Then, its contents is laid out into the containing box's resolved size normally, i.e. `contain:size` has no effect in this phase._"

Sure. That's a reasonable clarification.

> Perhaps also add a parenthetical clarification: "(including any pseudo-elements)" at the end.

This too seems like a useful clarification.

>  1. change "When calculating the size of the containing box, including when computing its intrinsic size, it must be treated as having no contents." to "When calculating the intrinsic size of the containing box, it must be treated as having no contents.".

That would not work:

If the size-contained grid is inside of a absolutely positioned element, or inside of a float which is itself inside a zero-width div, then when we try to figure out the size of the abspos or of the div, we look at the **intrinsic size** of its content (the size-positioned grid): the max-content size in the case of the abspos, the min-content size in the case of the float in a zero-width container. This is correctly captured by your phrasing (and also I believe by the one currently in the spec).

However, the effect is not limited to this, and your suggested edit drops the other part:

If a size-contained grid is an ordinary block-level child of an ordinary div, the grid must still be sized as if it were empty, then laid out in that size. For a grid with no gap, fixed-sized track, etc, than would mean a width equal to that of its block container, and a height of 0. This is not calculating the **intrinsic size** of the grid. It is calculating its **size**.

> My understanding is that size containment is only supposed to prevent the intrinsic size of an element from depending on the size of the contents. But in this case, the existence of implicit tracks wouldn't depend on the layout of the children, only on their computed style.
>
> So I'm not opposed to avoid creating implicit tracks with size containment, but I'm not convinced that's necessary.

When you change the style of the children of the grid to display:none, or to turn them into abspos, or to use grid-row or grid-column to explicitly place them into implicit tracks, you can affect the number of tracks implicit tracks, and if they have a non-zero fixed min track sizing function, this in turn affects the size of the grid even if you ignore them when sizing the tracks. But since the goal of size containment is that whatever happens inside the element has no effect on the size of the element, this cannot be allowed.

**Comment by MatsPalmgren** [comment]
> the grid must still be sized as if it were empty, then laid out in that size

Well, a box either intrinsically sized, or has a definite size from the style properties, or stretch fit into a definite available size. [There are no other cases](https://drafts.csswg.org/css-sizing-3/#auto-box-sizes), hence the part I removed is unnecessary and just adds confusion. Obviously, when the size is definite then whether or not it should be consider to have no contents when calculating that size is irrelevant. So the only case that remains is intrinsic sizing.

Anyway, I believe we agree on the behavior, just not which wording to best describe it.

> But since the goal of size containment is that whatever happens inside the element has no effect on the size of the element, this cannot be allowed.

OK, so you're advocating for option C (https://github.com/w3c/csswg-drafts/issues/4931#issuecomment-615144446) then, if I understand you correctly?

**Comment by frivoal** [comment]
Hmm. I accidentally let this sleep for a little too long.

In summary, at this point, I think we have narrowed it down to two possible approaches:

α) Clarification of the current spec, to stay true to its initial intent, which would have you size the element as if it was empty, including explicit grid tracks sized as if they had no element placed in them, and not including any implicit track since they wouldn't be created in the absence of content; then in a second phase, perform layout of the content within the now fixed-size box normally. Specific wording to be tweaked, but it's in the spirit of ["solution C"](https://github.com/w3c/csswg-drafts/issues/4931#issuecomment-615144446).

β) Make contain:size more aggressive, and actually suppress *explicit* grid tracks as well during the sizing step (but still lay everything normally in phase 2).
This isn't really necessary for the purposes of contain:size itself, but if we want to use it as a building block for some kind of container queries thing, that could be useful. Exactly how container queries is going to work (if it ever does) is unclear, but logically, in order to avoid circular definitions, you would not be able to set conditionally by query any properties that can affect the size of the object you're querying. Which means that grid-template can't be something you could change in response to a query about the size of the grid container, unless we made sure that contain:size suppresses its effect on the container’s size.

If we apply the same principles to multicol:
* the result under α would be that if both column-count and column-width are set, you size as having that many columns of that size (with colum-gaps inbetween); if only column-width is set, you size for 1 column of that size; and if only column-count is set, you size as n-1 times the size of the column gap.. 
* the result under β would be that you always size a multicol as an ordinary block.

Typically, size containment isn't used without manually providing a size: For out-of-flow elements, size containment isn't useful, and for in-flow elements, size containment without providing a size results in jumbled / overlapping layouts. So for most uses, the difference doesn't actually matter.

The one case where it would matter is 1D size containment. Having the inline dimension of a block level element be size-contained, without specifying a size manually, would be useful, and could very well serve as a basis for container queries. This is discussed in https://github.com/w3c/csswg-drafts/issues/1031 

That said, as discussed in that issue, it's not clear yet whether whether 1D size-containment + container queries is going to be possible at all.

Unless I am missing something about the usefulness of β with 2D size-containment, I think this leaves us with a few paths forward:
1. Decide now to do β anyway. It might not be useful if 1D size-containment + container queries turns out to be impossible, but the downside (other than spec and implementation churn) is relatively minor: size-containment without manual size becomes even worse at sizing things, but it was bad in the first place anyway.
2. Decide now to do α for 2D containment, but keep the question open for 1D containment, if we ever get it to work.
3. Punt until https://github.com/w3c/csswg-drafts/issues/1031 is resolved one way or another.

**Comment by frivoal** [comment]
I've made an exploratory pull request in https://github.com/w3c/csswg-drafts/pull/5596 that (attempts to) implement option α, and can be turned into β with the addition of a single sentence.

**Comment by frivoal** [comment]
This was resolved by merging #5643. The change is applied directly in level 2, and the corresponding change is marked up as a [candidate correction in level 1](https://drafts.csswg.org/css-contain-1/#c1) (to be published soon). Closing.
